core/utils/histogram.h

```diff
@@ -1,55 +1,180 @@
+  }
+
+  // Return the average value in the histogram.
+  // NOTE: This function s destructive, i.e. it turns the histogram into a
+  // cummulative histogram. Any calls to insert() after calls to this function
+  // will result in a call to reset().
+  T avg() {
+    if (!count_) {
+      summarize();
+    }
+    if (count_) {
+      return total_ / count_;
+    }
+    return 0;
   }
-}
-void print_hist(struct histogram* hist);
-struct histogram* combine_histograms(struct histogram* a, struct histogram* b);
 
-// Dump summary statistics for "hist" into "summary". This is destructive.
-void summarize_hist(struct histogram* hist, struct hist_summary *summary);
-void print_summary(struct hist_summary* s);
+  // Return the sum of the frequencies of each bucket.
+  // NOTE: This function s destructive, i.e. it turns the histogram into a
+  // cummulative histogram. Any calls to insert() after calls to this function
+  // will result in a call to reset().
+  size_t count() {
+    if (!count_) {
+      summarize();
+    }
+    return count_;
+  }
+
+  // Return the total of values inserted into the histogram, i.e. the sum of
+  // (i+1) * bucket_width_ * count(b_i) for each bucket b_i
+  // NOTE: This function s destructive, i.e. it turns the histogram into a
+  // cummulative histogram. Any calls to insert() after calls to this function
+  // will result in a call to reset().
+  T total() {
+    if (!count_) {
+      summarize();
+    }
+    return total_;
+  }
 
-// It's a good idea to call this after summarize_hist().
-void reset_hist(struct histogram *hist);
+  // Return the largest bucket b_i for which the cummulative frequency is less
+  // than p % of the total. p should be in [0, 100].
+  // NOTE: This function s destructive, i.e. it turns the histogram into a
+  // cummulative histogram. Any calls to insert() after calls to this function
+  // will result in a call to reset().
+  T percentile(double p) {
+    p /= 100.0;
+    if (!count_) {
+      summarize();
+    }
+
+    T ret = 0;
+    size_t p_count = (size_t)(p * count_);
+    for (size_t i = 0; i < max_bucket_; i++) {
+      if (buckets_[i] < p_count) {
+        ret = i + 1;
+      }
+    }
+    return ret * bucket_width_;
+  }
+
+  // Zero out the histogram.
+  void reset() {
+    count_ = 0;
+    total_ = 0;
+    min_bucket_ = 0;
+    max_bucket_ = 0;
+    above_threshold_ = 0;
+    memset(buckets_, 0, num_buckets_ * sizeof(size_t));
+  }
+
+ private:
+  // TODO(melvin): add support for logarithmic binning
+  size_t *bucket(T x) { return buckets_ + (uintptr_t)(x / bucket_width_); }
+
+  // Convert the histogram into a cummulative histogram. Called by min(), max(),
+  // avg(), count(), total() and percentile().
+  void summarize();
+
+  size_t num_buckets_;
+  T bucket_width_;
+  T threshold_;
+  size_t *buckets_;
+  size_t above_threshold_;
+  size_t count_;
+  T total_;
+  size_t min_bucket_;
+  size_t max_bucket_;
+};
+
+template <typename T>
+void Histogram<T>::summarize() {
```

Comment:
It's a bit longer than the other methods. Looks weird all by itself on the outside, though. Will inline.